### PR TITLE
libyang python APIs

### PR DESCRIPTION
**Pull request by li-pingmao**
**- What I did**
        libyang Python APIs:
          - python APIs based on libyang
          - functions to load/merge yang models and Yang data files
          - function to validate data trees based on Yang models
          - functions to merge yang data files/trees
          - add/set/delete node in schema and data trees
          - find data/schema nodes from xpath from the Yang data/schema tree in memory 
          - find dependencies
          - dump the data tree in json/xml 

**- How I did it**
     -  based on libyang python2 modules which is built by libyang
     -  add functions to load/validate Yang models and Yang data trees
     -  Add functions to 
    
**- How to verify it**
    -  pytests while build sonic-yang-mgmt python wheel package
    -  create sample Yang models for sonic ACL/interface/port/clan
    -  sample Yang data files
    -  test file with test cases,  it includes categories to test, xpath and expected results for testing Yang data/schema tree wrapper functions

pytests while build target/python-wheels/sonic_yang_mgmt-1.0-py2-none-any.whl:

============================= test session starts ==============================
platform linux2 -- Python 2.7.13, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /sonic/src/sonic-yang-mgmt, inifile:
plugins: cov-2.4.0
collected 18 items

tests/test_sonic_yang_mgmt.py .
tests/libyang-python-tests/test_sonic_yang.py .................

========================== 18 passed in 0.21 seconds ===========================

**- Description for the changelog**

Note:    Due to dependency on PR#3861 (https://github.com/Azure/sonic-buildimage/pull/3861), this PR has 2 commits, please review the 2nd commit only(the first commit in this PR is a cherry-pick of PR#3861).
      

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: rules/sonic-yang-mgmt-py2.mk [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# sonic-yang-mgmt python2 wheel
+
+SONIC_YANG_MGMT_PY2 = sonic_yang_mgmt-1.0-py2-none-any.whl
```
Comment (renukamanavalan) [inline rules/sonic-yang-mgmt-py2.mk]: Let us not tag it PY2.  We have to move to Python3 in 201911 image. So at least call .py and when system default switches to python3, we will use it transparently.

File: src/sonic-yang-mgmt/setup.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        if (errno):
+            exit(errno)
+
+setup(
+    cmdclass={
+        'build_py': pkgBuild,
+    },
+    author="lnos-coders",
+    author_email='lnos-coders@linkedin.com',
+    python_requires='>=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*',
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/setup.py]: Same comment as before. Can we switch to >= 3.5

File: src/sonic-yang-mgmt/sonic_yang.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+            self.ctx = ly.Context(yang_dir)
+
+        try:
+            self.load_schema_module_list(yang_files)
+            if len(data_files) == 0:
+                return (self.ctx, self.root)
+
+            self.load_data_file(data_files[0])
+
+            for i in range(2, len(data_files)):
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/sonic_yang.py]: What about data_files[1]

File: src/sonic-yang-mgmt/sonic_yang.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+        return path
+
+    """
+    new_node(): create a new data node in the data tree
+    input:
+           xpath: xpath of the new node
+           value: value of the new node
+    returns:  new Data_Node object if success,  Exception if falied
+    """
+    def new_node(self, xpath, value):
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/sonic_yang.py]: Need a standard.
Here we load schema into 'yang.Context' and data into root Node, where each node has associated schema node. 
Can we standardize as

- Node implies data node

- Schema node is explicitly called out as "snode"

- APIs that deal with data node, simply says node (e.g. set_node_value, find_nodes, ...)

- APIs that refer/return schema node, explicitly mention schema (e.g. find_schema_node)

Can we drop dnode, data_nodes, ...

File: src/sonic-yang-mgmt/sonic_yang.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+                    value = subtype.value_str()
+                    return value
+            return output
+
+    """
+    set the value of a node in the data tree
+    input:    xpath of the data node
+    returns:  Exception if failed
+    """
+    def set_dnode_value(self, data_xpath, value):
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/sonic_yang.py]: set_node_value ?

File: src/sonic-yang-mgmt/sonic_yang.py [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+        except Exception as e:
+            print("set data node value failed for xpath: " + str(data_xpath))
+            self.fail(e)
+
+    """
+    find_data_nodes(): find the set of nodes for the xpath
+    input:    xpath of the data node
+    returns:  list of xpath of the dataset
+    """
+    def find_data_nodes(self, data_xpath):
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/sonic_yang.py]: find_nodes?

File: src/sonic-yang-mgmt/tests/libyang-python-tests/sample-yang-models/sonic-acl.yang [hunk 7]
```diff
@@ -0,0 +1,274 @@
+module sonic-acl {
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/tests/libyang-python-tests/sample-yang-models/sonic-acl.yang]: Why do we have a sample for schema ?

In other words, why it is not checked in as proper schema ?

Comment (li-pingmao) [reply src/sonic-yang-mgmt/tests/libyang-python-tests/sample-yang-models/sonic-acl.yang]: - The yang models are in separate PR.   
- These sample models are used for testing the libyang  python APIs,   these APIs should work for any yang models. 
-  yang models may change/update, it's better to separate the yang models and the sample Yang models for testing libyang python APIs.

File: src/sonic-yang-mgmt/sonic_yang.py (hunk 6 above)
Comment (li-pingmao) [reply src/sonic-yang-mgmt/sonic_yang.py]: libyang has data tree and schema tree.  This API finds data nodes, there is a separate API for find the schema node.

File: src/sonic-yang-mgmt/tests/libyang-python-tests/sample-yang-models/sonic-acl.yang [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+	yang-version 1.1;
+
+	namespace "http://github.com/Azure/sonic-acl";
```
Comment (renukamanavalan) [inline src/sonic-yang-mgmt/tests/libyang-python-tests/sample-yang-models/sonic-acl.yang]: rename sonic-acl.yang to something like test1-acl.yang or ...

File: src/sonic-yang-mgmt/sonic_yang.py [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+        try:
+            module = self.ctx.get_module(str(module_name))
+        except Exception as e:
+            print("Cound not get module: " + str(module_name))
+            self.fail(e)
+        else:
+            if (module is not None):
+                if (format == "XML"):
+                    #libyang bug with format
+                    result = module.print_mem(ly.LYD_JSON, ly.LYP_FORMAT)
```
[the remark this answers is not in the document]
Comment (li-pingmao) [reply src/sonic-yang-mgmt/sonic_yang.py]: A bug in libyang

**Comment by praveen-li** [comment]
@li-pingmao : Kindly close this PR, by giving reference to PR 3861. Thx

**Comment by li-pingmao** [comment]
This PR was pushed as part of PR 3861.

**Pull request closed without merging**